### PR TITLE
Updated explain command

### PR DESCRIPTION
**Pull request by matentzn**
Various updates

Resolves [#756, #329, #686]

- [X] `docs/` have been added/updated
- [X] tests have been added/updated
- [ ] `mvn verify` says all tests pass <- cant run 
image

- [ ] `mvn site` says all JavaDocs correct <- cant run (see #777)
- [X] `CHANGELOG.md` has been updated

This pull request contains a number of changes to the explain command. The original command behaviour (which I refer to as `entailment` explanations) are untouched. See updated documentation for details. 

### Summary of changes:

1. We can now generate explanations for 
    - inconsistent ontologies
    - unsatisfiable classes (all at once, root unsats, random subset)
2. Markdown exports get a summary which shows the most frequently used axioms across all found ontologies. This will help with debugging by focusing the attention on likely culprits.
3. Added some unit tests to `explain` and updated docs.

To achieve the above, there are now two new options:

- `-M\--mode`: allows switching between `entailment` (default), `inconsistency` and `unsatisfiablity`.
- `-u\--unsatisfiable`: allows various fine grained options for generating explanations for unsatisfiable classes:
    - `all` (generate explanations for all unsats)
    - `root` (generate explanations for root unsats)
    - `random:n` subset (generate explanations for `n` random unsatisfiable classes.

```
Example entailment summary (under the normal explanations in the markdown export):
```
## [uvular muscle](http://purl.obolibrary.org/obo/UBERON_0010235) SubClassOf [Nothing](http://www.w3.org/2002/07/owl#Nothing) ##

  - [uvular muscle](http://purl.obolibrary.org/obo/UBERON_0010235) SubClassOf [palatal muscle](http://purl.obolibrary.org/obo/UBERON_0003682)
  - [palatal muscle](http://purl.obolibrary.org/obo/UBERON_0003682) DisjointWith [uvular muscle](http://purl.obolibrary.org/obo/UBERON_0010235)

## [epimysium](http://purl.obolibrary.org/obo/UBERON_0011899) SubClassOf [Nothing](http://www.w3.org/2002/07/owl#Nothing) ##

  - [epimysium](http://purl.obolibrary.org/obo/UBERON_0011899) SubClassOf [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822)
    - [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822) SubClassOf [irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011821)
      - [irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011821) SubClassOf [connective tissue](http://purl.obolibrary.org/obo/UBERON_0002384)
        - [connective tissue](http://purl.obolibrary.org/obo/UBERON_0002384) SubClassOf [tissue](http://purl.obolibrary.org/obo/UBERON_0000479)
          - [tissue](http://purl.obolibrary.org/obo/UBERON_0000479) SubClassOf [multicellular anatomical structure](http://purl.obolibrary.org/obo/UBERON_0010000)
    - [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822) SubClassOf [dense connective tissue](http://purl.obolibrary.org/obo/UBERON_0011823)
  -  DisjointClasses: [multicellular anatomical structure](http://purl.obolibrary.org/obo/UBERON_0010000), [dense connective tissue](http://purl.obolibrary.org/obo/UBERON_0011823), [developing mesenchymal structure](http://purl.obolibrary.org/obo/UBERON_0017650)

## [skeletal muscle tissue](http://purl.obolibrary.org/obo/UBERON_0001134) SubClassOf [Nothing](http://www.w3.org/2002/07/owl#Nothing) ##

  - [skeletal muscle tissue](http://purl.obolibrary.org/obo/UBERON_0001134) SubClassOf [part of](http://purl.obolibrary.org/obo/BFO_0000050) some [skeletal muscle organ](http://purl.obolibrary.org/obo/UBERON_0014892)
    - [skeletal muscle organ](http://purl.obolibrary.org/obo/UBERON_0014892) EquivalentTo [muscle organ](http://purl.obolibrary.org/obo/UBERON_0001630) and ([develops_from](http://purl.obolibrary.org/obo/RO_0002202) some [myotome](http://purl.obolibrary.org/obo/UBERON_0003082)) and ([surrounded_by](http://purl.obolibrary.org/obo/RO_0002219) some [epimysium](http://purl.obolibrary.org/obo/UBERON_0011899)) and ([composed primarily of](http://purl.obolibrary.org/obo/RO_0002473) some [skeletal muscle tissue](http://purl.obolibrary.org/obo/UBERON_0001134)) and ([synapsed by](http://purl.obolibrary.org/obo/uberon/core#synapsed_by) some [motor neuron](http://purl.obolibrary.org/obo/CL_0000100))
      - [epimysium](http://purl.obolibrary.org/obo/UBERON_0011899) SubClassOf [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822)
        - [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822) SubClassOf [irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011821)
          - [irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011821) SubClassOf [connective tissue](http://purl.obolibrary.org/obo/UBERON_0002384)
            - [connective tissue](http://purl.obolibrary.org/obo/UBERON_0002384) SubClassOf [tissue](http://purl.obolibrary.org/obo/UBERON_0000479)
              - [tissue](http://purl.obolibrary.org/obo/UBERON_0000479) SubClassOf [multicellular anatomical structure](http://purl.obolibrary.org/obo/UBERON_0010000)
        - [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822) SubClassOf [dense connective tissue](http://purl.obolibrary.org/obo/UBERON_0011823)
  -  DisjointClasses: [multicellular anatomical structure](http://purl.obolibrary.org/obo/UBERON_0010000), [dense connective tissue](http://purl.obolibrary.org/obo/UBERON_0011823), [developing mesenchymal structure](http://purl.obolibrary.org/obo/UBERON_0017650)

## [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822) SubClassOf [Nothing](http://www.w3.org/2002/07/owl#Nothing) ##

  - [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822) SubClassOf [irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011821)
    - [irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011821) SubClassOf [connective tissue](http://purl.obolibrary.org/obo/UBERON_0002384)
      - [connective tissue](http://purl.obolibrary.org/obo/UBERON_0002384) SubClassOf [tissue](http://purl.obolibrary.org/obo/UBERON_0000479)
        - [tissue](http://purl.obolibrary.org/obo/UBERON_0000479) SubClassOf [multicellular anatomical structure](http://purl.obolibrary.org/obo/UBERON_0010000)
  - [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822) SubClassOf [dense connective tissue](http://purl.obolibrary.org/obo/UBERON_0011823)
  -  DisjointClasses: [multicellular anatomical structure](http://purl.obolibrary.org/obo/UBERON_0010000), [dense connective tissue](http://purl.obolibrary.org/obo/UBERON_0011823), [developing mesenchymal structure](http://purl.obolibrary.org/obo/UBERON_0017650)

## [skeletal muscle organ](http://purl.obolibrary.org/obo/UBERON_0014892) SubClassOf [Nothing](http://www.w3.org/2002/07/owl#Nothing) ##

  - [skeletal muscle organ](http://purl.obolibrary.org/obo/UBERON_0014892) EquivalentTo [muscle organ](http://purl.obolibrary.org/obo/UBERON_0001630) and ([develops_from](http://purl.obolibrary.org/obo/RO_0002202) some [myotome](http://purl.obolibrary.org/obo/UBERON_0003082)) and ([surrounded_by](http://purl.obolibrary.org/obo/RO_0002219) some [epimysium](http://purl.obolibrary.org/obo/UBERON_0011899)) and ([composed primarily of](http://purl.obolibrary.org/obo/RO_0002473) some [skeletal muscle tissue](http://purl.obolibrary.org/obo/UBERON_0001134)) and ([synapsed by](http://purl.obolibrary.org/obo/uberon/core#synapsed_by) some [motor neuron](http://purl.obolibrary.org/obo/CL_0000100))
    - [epimysium](http://purl.obolibrary.org/obo/UBERON_0011899) SubClassOf [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822)
      - [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822) SubClassOf [irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011821)
        - [irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011821) SubClassOf [connective tissue](http://purl.obolibrary.org/obo/UBERON_0002384)
          - [connective tissue](http://purl.obolibrary.org/obo/UBERON_0002384) SubClassOf [tissue](http://purl.obolibrary.org/obo/UBERON_0000479)
            - [tissue](http://purl.obolibrary.org/obo/UBERON_0000479) SubClassOf [multicellular anatomical structure](http://purl.obolibrary.org/obo/UBERON_0010000)
      - [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822) SubClassOf [dense connective tissue](http://purl.obolibrary.org/obo/UBERON_0011823)
  -  DisjointClasses: [multicellular anatomical structure](http://purl.obolibrary.org/obo/UBERON_0010000), [dense connective tissue](http://purl.obolibrary.org/obo/UBERON_0011823), [developing mesenchymal structure](http://purl.obolibrary.org/obo/UBERON_0017650)

## [palatal muscle](http://purl.obolibrary.org/obo/UBERON_0003682) SubClassOf [Nothing](http://www.w3.org/2002/07/owl#Nothing) ##

  - [palatal muscle](http://purl.obolibrary.org/obo/UBERON_0003682) EquivalentTo [skeletal muscle organ](http://purl.obolibrary.org/obo/UBERON_0014892) and ([part of](http://purl.obolibrary.org/obo/BFO_0000050) some [soft palate](http://purl.obolibrary.org/obo/UBERON_0001733))
    - [skeletal muscle organ](http://purl.obolibrary.org/obo/UBERON_0014892) EquivalentTo [muscle organ](http://purl.obolibrary.org/obo/UBERON_0001630) and ([develops_from](http://purl.obolibrary.org/obo/RO_0002202) some [myotome](http://purl.obolibrary.org/obo/UBERON_0003082)) and ([surrounded_by](http://purl.obolibrary.org/obo/RO_0002219) some [epimysium](http://purl.obolibrary.org/obo/UBERON_0011899)) and ([composed primarily of](http://purl.obolibrary.org/obo/RO_0002473) some [skeletal muscle tissue](http://purl.obolibrary.org/obo/UBERON_0001134)) and ([synapsed by](http://purl.obolibrary.org/obo/uberon/core#synapsed_by) some [motor neuron](http://purl.obolibrary.org/obo/CL_0000100))
      - [epimysium](http://purl.obolibrary.org/obo/UBERON_0011899) SubClassOf [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822)
        - [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822) SubClassOf [irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011821)
          - [irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011821) SubClassOf [connective tissue](http://purl.obolibrary.org/obo/UBERON_0002384)
            - [connective tissue](http://purl.obolibrary.org/obo/UBERON_0002384) SubClassOf [tissue](http://purl.obolibrary.org/obo/UBERON_0000479)
              - [tissue](http://purl.obolibrary.org/obo/UBERON_0000479) SubClassOf [multicellular anatomical structure](http://purl.obolibrary.org/obo/UBERON_0010000)
        - [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822) SubClassOf [dense connective tissue](http://purl.obolibrary.org/obo/UBERON_0011823)
  -  DisjointClasses: [multicellular anatomical structure](http://purl.obolibrary.org/obo/UBERON_0010000), [dense connective tissue](http://purl.obolibrary.org/obo/UBERON_0011823), [developing mesenchymal structure](http://purl.obolibrary.org/obo/UBERON_0017650)

# Axiom Impact 
## Axioms used 5 times
- [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822) SubClassOf [dense connective tissue](http://purl.obolibrary.org/obo/UBERON_0011823) [uberon-base.owl]
- [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822) SubClassOf [irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011821) [uberon-base.owl]
-  DisjointClasses: [multicellular anatomical structure](http://purl.obolibrary.org/obo/UBERON_0010000), [dense connective tissue](http://purl.obolibrary.org/obo/UBERON_0011823), [developing mesenchymal structure](http://purl.obolibrary.org/obo/UBERON_0017650) [uberon-base.owl]
- [tissue](http://purl.obolibrary.org/obo/UBERON_0000479) SubClassOf [multicellular anatomical structure](http://purl.obolibrary.org/obo/UBERON_0010000) [uberon-base.owl]
- [irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011821) SubClassOf [connective tissue](http://purl.obolibrary.org/obo/UBERON_0002384) [uberon-base.owl]
- [connective tissue](http://purl.obolibrary.org/obo/UBERON_0002384) SubClassOf [tissue](http://purl.obolibrary.org/obo/UBERON_0000479) [uberon-base.owl]

## Axioms used 4 times
- [epimysium](http://purl.obolibrary.org/obo/UBERON_0011899) SubClassOf [dense irregular connective tissue](http://purl.obolibrary.org/obo/UBERON_0011822) [uberon-base.owl]

## Axioms used 3 times
- [skeletal muscle organ](http://purl.obolibrary.org/obo/UBERON_0014892) EquivalentTo [muscle organ](http://purl.obolibrary.org/obo/UBERON_0001630) and ([develops_from](http://purl.obolibrary.org/obo/RO_0002202) some [myotome](http://purl.obolibrary.org/obo/UBERON_0003082)) and ([surrounded_by](http://purl.obolibrary.org/obo/RO_0002219) some [epimysium](http://purl.obolibrary.org/obo/UBERON_0011899)) and ([composed primarily of](http://purl.obolibrary.org/obo/RO_0002473) some [skeletal muscle tissue](http://purl.obolibrary.org/obo/UBERON_0001134)) and ([synapsed by](http://purl.obolibrary.org/obo/uberon/core#synapsed_by) some [motor neuron](http://purl.obolibrary.org/obo/CL_0000100)) [uberon-base.owl]

## Axioms used 1 times
- [uvular muscle](http://purl.obolibrary.org/obo/UBERON_0010235) SubClassOf [palatal muscle](http://purl.obolibrary.org/obo/UBERON_0003682) [uberon-base.owl]
- [palatal muscle](http://purl.obolibrary.org/obo/UBERON_0003682) EquivalentTo [skeletal muscle organ](http://purl.obolibrary.org/obo/UBERON_0014892) and ([part of](http://purl.obolibrary.org/obo/BFO_0000050) some [soft palate](http://purl.obolibrary.org/obo/UBERON_0001733)) [uberon-base.owl]
- [skeletal muscle tissue](http://purl.obolibrary.org/obo/UBERON_0001134) SubClassOf [part of](http://purl.obolibrary.org/obo/BFO_0000050) some [skeletal muscle organ](http://purl.obolibrary.org/obo/UBERON_0014892) [uberon-base.owl]
- [palatal muscle](http://purl.obolibrary.org/obo/UBERON_0003682) DisjointWith [uvular muscle](http://purl.obolibrary.org/obo/UBERON_0010235) [uberon-base.owl]

# Ontologies used: 
- uberon-base.owl (http://purl.obolibrary.org/obo/uberon/uberon-base.owl)

### CODE REVIEW

**Comment by matentzn** [comment]
@dosumis @cmungall FYI

**Comment by matentzn** [comment]
I have just added a way to see which ontologies an axiom comes from in the markdown overview. Note this only works with real imports, not with the `source_ontology` annotation.

**Comment by matentzn** [comment]
I walked @cmungall through this PR and he is happy

**Review by balhoff — APPROVED** [review]
This is great!

> I have just added a way to see which ontologies an axiom comes from in the markdown overview.

Does this also work with the previous entailment mode?

**Comment by matentzn** [comment]
It does! But only in the summary at the bottom. So for a 1 explanation scenario it will be redundant, but yes, you get that information!

File: robot-command/src/main/java/org/obolibrary/robot/ExplainCommand.java [hunk 1]
```diff
@@ -31,7 +31,25 @@
   private static final Logger logger = LoggerFactory.getLogger(ExplainCommand.class);
 
   private static final String NS = "explain#";
+
+  private static final List<String> legal_modes =
+      Arrays.asList("entailment", "inconsistency", "unsatisfiability");
+  OWLDataFactory df = OWLManager.getOWLDataFactory();
+
   private static final String maxTypeError = NS + "MAX TYPE ERROR --max ('%s') must be an integer";
+  private static final String inconsistentOntologyError =
```
Comment (beckyjackson) [inline robot-command/src/main/java/org/obolibrary/robot/ExplainCommand.java:40]: This is never used

File: robot-core/src/main/java/org/obolibrary/robot/ExplainOperation.java [hunk 2]
```diff
@@ -112,6 +304,38 @@ private static String renderTree(Tree<OWLAxiom> tree, OWLObjectRenderer renderer
     return builder.toString();
   }
 
+  private static String renderAxiomWithImpact(
+      List<OWLAxiom> axioms,
+      int impact,
+      OWLObjectRenderer renderer,
+      Map<OWLAxiom, Set<String>> associatedOntologyIds) {
+    StringBuilder builder = new StringBuilder();
+    builder.append("## Axioms used " + impact + " times" + "\n");
```
Comment (beckyjackson) [inline robot-core/src/main/java/org/obolibrary/robot/ExplainOperation.java:313]: String concats in StringBuilders should be chained builder calls (better performance):
```
builder.append("## Axiom used ").append(impact).append(" times\n");
```

File: robot-core/src/main/java/org/obolibrary/robot/ExplainOperation.java [hunk 3]
```diff
@@ -56,6 +58,117 @@
...
+   * @return a set of explanations
+   */
+  public static Set<Explanation<OWLAxiom>> explainMostGeneralUnsatisfiableClasses(
+      OWLOntology ontology, OWLReasoner reasoner, OWLReasonerFactory reasonerFactory, int max) {
+    List<OWLClass> unsatisfiable_classes =
+        new ArrayList<>(getMostGeneralUnsatisfiableClasses(reasoner, ontology));
+    return getUnsatExplanationsForClasses(ontology, reasonerFactory, max, unsatisfiable_classes);
+  }
+
+  private static Set<OWLClass> getMostGeneralUnsatisfiableClasses(
```
Comment (beckyjackson) [inline robot-core/src/main/java/org/obolibrary/robot/ExplainOperation.java:154]: Please move this to be with the other private methods (order public methods -> private methods)

**Review by beckyjackson — COMMENTED** [review]
Overall code looks good - see comments.

**Comment by matentzn** [comment]
I converted this to draft to incorporate all of @beckyjackson feedback from the `measure` command first before giving this up for review.

https://github.com/ontodev/robot/pull/778

**Comment by jamesaoverton** [comment]
I tried updating the `japicmp` pluging in `robot-core/pom.xml` to the latest version 0.15.2. Then the integration tests passed under Java 8. Under Java 15 (in Docker) I got one failure involving the order of prefixes in one test file. I don't think that's a problem with the code here. I also made a general issue about updating plugins #798.

**Review by jamesaoverton — APPROVED** [review]

**Comment by jamesaoverton** [comment]
Sorry for the delay @matentzn.

@beckyjackson Please merge when ready.

**Pull request merged**